docs/2.10.0/docs/upgrade/smart-contract-upgrades.rst

```diff
@@ -2264,6 +2283,114 @@ even when it is not a valid upgrade from one version to the next.
+
+When considering upgrading, we regard each Daml application as composed of:
+
+- **On-ledger components**: The Daml code running the on-ledger logic (i.e. the DARs uploaded
+  to all participant nodes interacting with the app)
+
+- **Off-ledger components** interacting with the ledger via the Ledger API or other supported
+  Canton Ledger API clients (JSON API or PQS).
+  These are typically Java or TypeScript services implementing off-ledger business logic.
+
+Zero-Downtime Upgrades
+~~~~~~~~~~~~~~~~~~~~~~
+
+Upgrading an application without operational downtime can be achieved by designing the above-mentioned components to allow:
+
+- **Asynchronous rollout**: Operators deploy updated software at their own pace,
+  similar to established CI/CD practices in micro-service environments.
+
+- **Synchronous switch-over**: All components switch to using updated Daml code at the same time.
+
+**Designing for asynchronous upgrade roll-outs**
+
+To allow for asynchronous roll-outs, off-ledger components should:
+
+- **use package names in Ledger API requests**:
+  App components interacting with the Ledger API should use `by-package-name template IDs`
+  instead of `by-package-id template IDs` in all their command submissions and queries.
+  This allows:
+
+    - Ledger API server-side version selection of the package preference for command submissions.
+
+    - reading all versions of the templates in queries, even newer versions
+      than the one the component was developed against.
+
+- **handle missing optional fields**:
+  App components reading from the Ledger API or Ledger API clients
+  must be prepared to handle missing optional fields in records,
+  including those in the initial package.
+  The TypeScript and Java codegens for reading Daml values do so by default.
+
+- **use exhaustive package preference**: on each command submission, the ``package_id_selection_preference`` is set ensuring that:
+
+    - For every package used in the command and for every package used for the all possible interface instances,
+      there is an association between the package name and the package ID of a precise package included in the package preference.
+
+    - Within an application, all submissions use the same package preference.
+
+.. note::
+    Following these recommendations decouples the application's behavior from the DAR uploads
+    (see :ref:`the note regarding Package Preference <package-preference>`),
+    as package resolution is fully determined by the explicit
+    and exhaustive package preference provided in command submissions.
+
+**Operational steps for upgrading**
+
+Once a new Daml package version of the application is ready, define an operational window
+for allowing the asynchronous rollout of the updated components.
+During this window, Canton node and off-ledger app operators should:
+
+- Upload the upgraded DARs to the participant nodes
+
+- Roll-out the updated off-ledger components
```

Comment:
It is a shame that this has to happen during the, possibly short operational window. There may be several apps each of which needs approvals etc. Another way to do it could be to roll out an app that can work with either the old or new daml models and just need to be switched to use the new package-preference list during the operational window.

Reply:
> Another way to do it could be to roll out an app that can work with either the old or new daml models and just need to be switched to use the new package-preference list during the operational window.

So the idea is exactly this: The apps work in a backwards compatible manner. The operational window is there just for setting some deadline after which the new code is executed. Or you meant that we can even go without a deadline? That would really be tricky since it would mean that off-ledger services should have some sort of forward compatibility (old off-ledger services being able to seamlessly participate to the upgraded workflow). Happy to discuss further though and open another amendment PR if you feel strongly about this.